File: obs_replaykit/cleanup.py
"""in-menu cleanup for OBS ReplayKit installs."""

from __future__ import annotations

import os
import shutil
import subprocess
from pathlib import Path
from typing import Callable, Optional

from .config import OBS_CONFIG, PROGRAMDATA, PROGRAMFILES_OBS_DIR, REPLAYKIT_CONFIG, REPLAYKIT_SETUP_CACHE, REPLAYKIT_TRAY_PLUGIN_DIR, REPLAYKIT_USER_STATE_CACHE
from .obs import close_obs
from .scheduled_task import delete_elevation_task
from .sleep_override import remove_sleep_override
from .startup import configure_obs_startup
from .vbcable import uninstall_vbcable


LogFn = Optional[Callable[[str], None]]

_REPLAYKIT_RUNTIME_STATE_RELS = (
    Path("obs-replayKit/scripts/helper/clips_db.json"),
    Path("obs-replayKit/scripts/helper/clips_index.json"),
)


def _run_hidden(args: list[str], timeout: int = 30) -> subprocess.CompletedProcess:
    return subprocess.run(
        args,
        capture_output=True,
        text=True,
        timeout=timeout,
        creationflags=getattr(subprocess, "CREATE_NO_WINDOW", 0),
    )


def _current_parent_pid() -> int | None:
    try:
        result = _run_hidden([
            "powershell.exe",
            "-NoProfile",
            "-NonInteractive",
            "-Command",
            f"(Get-CimInstance Win32_Process -Filter \"ProcessId={os.getpid()}\").ParentProcessId",
        ])
    except (OSError, subprocess.TimeoutExpired):
        return None
    try:
        return int((result.stdout or "").strip())
    except ValueError:
        return None


def stop_obs_and_helpers(log: LogFn = None) -> bool:
    """close OBS and ReplayKit helper processes while keeping this setup process alive."""
    close_obs(log=log)
    keep = {os.getpid()}
    parent = _current_parent_pid()
    if parent:
        keep.add(parent)
    keep_csv = ",".join(str(pid) for pid in keep)
    script = rf"""
$keep = @({keep_csv})
Get-CimInstance Win32_Process |
  Where-Object {{ $_.Name -in @('OBSReplayKit.exe','OBSReplayKit-Encoder.exe') -and $keep -notcontains [int]$_.ProcessId }} |
  ForEach-Object {{
    try {{ Stop-Process -Id $_.ProcessId -Force -ErrorAction Stop }} catch {{ }}
  }}
"""
    try:
        _run_hidden(["powershell.exe", "-NoProfile", "-NonInteractive", "-Command", script])
    except (OSError, subprocess.TimeoutExpired) as exc:
        if log:
            log(f"warn: helper stop failed: {exc}")
        return False
    return True


def remove_replaykit_plugins(log: LogFn = None) -> bool:
    targets = (
        PROGRAMFILES_OBS_DIR / "obs-plugins" / "64bit" / "win-capture-audio.dll",
        PROGRAMFILES_OBS_DIR / "obs-plugins" / "64bit" / "win-capture-audio.pdb",
        PROGRAMFILES_OBS_DIR / "data" / "obs-plugins" / "win-capture-audio",
        PROGRAMFILES_OBS_DIR / "obs-plugins" / "64bit" / "input-overlay.dll",
        PROGRAMFILES_OBS_DIR / "obs-plugins" / "64bit" / "SDL2.dll",
        PROGRAMFILES_OBS_DIR / "data" / "obs-plugins" / "input-overlay",
        PROGRAMFILES_OBS_DIR / "obs-plugins" / "64bit" / "bongobs-cat.dll",
        PROGRAMFILES_OBS_DIR / "bin" / "64bit" / "Bango Cat",
        PROGRAMFILES_OBS_DIR / "data" / "obs-plugins" / "bongobs-cat",
        PROGRAMFILES_OBS_DIR / "obs-plugins" / "64bit" / "obs-composite-blur.dll",
        PROGRAMFILES_OBS_DIR / "obs-plugins" / "64bit" / "obs-composite-blur.pdb",
        PROGRAMFILES_OBS_DIR / "data" / "obs-plugins" / "obs-composite-blur",
        PROGRAMDATA / "obs-studio" / "plugins" / "obs-composite-blur",
        PROGRAMFILES_OBS_DIR / "obs-plugins" / "64bit" / "obs-shaderfilter.dll",
        PROGRAMFILES_OBS_DIR / "obs-plugins" / "64bit" / "obs-shaderfilter.pdb",
        PROGRAMFILES_OBS_DIR / "data" / "obs-plugins" / "obs-shaderfilter",
        REPLAYKIT_TRAY_PLUGIN_DIR,
    )
    ok = True
    for target in targets:
        if not target.exists():
            continue
        try:
            if target.is_dir():
                shutil.rmtree(target)
            else:
                target.unlink()
        except OSError as exc:
            ok = False
            if log:
                log(f"warn: could not remove {target}: {exc}")
    return ok


def remove_virtual_display_driver(log: LogFn = None) -> bool:
    script = r"""
$ErrorActionPreference = 'Continue'
$vddDevice = Get-PnpDevice -ErrorAction SilentlyContinue |
    Where-Object { $_.FriendlyName -eq 'Virtual Display Driver' } |
    Select-Object -First 1
if ($vddDevice) {
    Disable-PnpDevice -InstanceId $vddDevice.InstanceId -Confirm:$false -ErrorAction SilentlyContinue | Out-Null
    pnputil.exe /remove-device $vddDevice.InstanceId | Out-Null
}
$drivers = pnputil.exe /enum-drivers | Out-String
$matches = [regex]::Matches($drivers, "Published Name:\s+(oem\d+\.inf)\s+Original Name:\s+MttVDD\.inf", "IgnoreCase")
foreach ($m in $matches) {
    pnputil.exe /delete-driver $m.Groups[1].Value /uninstall /force | Out-Null
}
if (Test-Path "C:\IddSampleDriver") {
    Remove-Item -Recurse -Force "C:\IddSampleDriver" -ErrorAction SilentlyContinue
}
"""
    try:
        result = _run_hidden(["powershell.exe", "-NoProfile", "-NonInteractive", "-Command", script], timeout=60)
    except (OSError, subprocess.TimeoutExpired) as exc:
        if log:
            log(f"warn: virtual display cleanup failed: {exc}")
        return False
    if result.returncode != 0 and log:
        err = (result.stderr or result.stdout or "").strip().splitlines()
        log(f"warn: virtual display cleanup returned {result.returncode}: {err[0] if err else ''}")
    return result.returncode == 0


def wipe_obs_config(log: LogFn = None) -> bool:
    if not REPLAYKIT_CONFIG.exists():
        return True
    try:
        shutil.rmtree(REPLAYKIT_CONFIG)
        return True
    except OSError as exc:
        if log:
            log(f"warn: could not wipe {REPLAYKIT_CONFIG}: {exc}")
        return False


def save_user_settings(log: LogFn = None) -> bool:
    ok = True
    try:
        from .prefs import PREFS_FILE, load_prefs

        prefs = load_prefs()
        prefs.save()
        if log:
            log(f"ReplayKit user settings kept -> {PREFS_FILE}")
    except Exception as exc:
        ok = False
        if log:
            log(f"warn: could not keep ReplayKit user settings: {exc}")

    try:
        REPLAYKIT_USER_STATE_CACHE.mkdir(parents=True, exist_ok=True)
        kept = 0
        for rel in _REPLAYKIT_RUNTIME_STATE_RELS:
            source = OBS_CONFIG / rel
            if not source.is_file():
                continue
            target = REPLAYKIT_USER_STATE_CACHE / rel.name
            shutil.copy2(source, target)
            kept += 1
        if kept and log:
            log(f"ReplayKit clip state kept -> {REPLAYKIT_USER_STATE_CACHE}")
    except OSError as exc:
        ok = False
        if log:
            log(f"warn: could not keep ReplayKit clip state: {exc}")
    return ok


def remove_user_settings(log: LogFn = None) -> bool:
    try:
        from .prefs import PREFS_FILE
    except Exception:
        PREFS_FILE = None

    targets: list[Path] = []
    if PREFS_FILE is not None:
        targets.append(PREFS_FILE)
    targets.append(REPLAYKIT_SETUP_CACHE / "prefs.json")
    targets.append(REPLAYKIT_SETUP_CACHE / "clips_state.json")

    ok = True
    seen: set[str] = set()
    for target in targets:
        key = str(target.resolve(strict=False)).lower()
        if key in seen:
            continue
        seen.add(key)
        if not target.exists():
            continue
        try:
            target.unlink()
        except OSError as exc:
            ok = False
            if log:
                log(f"warn: could not remove {target}: {exc}")
    if REPLAYKIT_USER_STATE_CACHE.exists():
        try:
            shutil.rmtree(REPLAYKIT_USER_STATE_CACHE)
        except OSError as exc:
            ok = False
            if log:
                log(f"warn: could not remove {REPLAYKIT_USER_STATE_CACHE}: {exc}")
    return ok


def run_cleanup(progress, keep_user_settings: bool = True) -> list[str]:
    steps: list[tuple[str, str, Callable[[], object]]] = []

    def add(title: str, detail: str, action: Callable[[], object]) -> None:
        steps.append((title, detail, action))

    add("Close OBS", "Stops OBS and ReplayKit helpers, but keeps this setup window alive.", lambda: stop_obs_and_helpers(log=progress.log))
    add("Remove launch permission", "Deletes the ReplayKit scheduled task.", lambda: delete_elevation_task(log=progress.log))
    add("Remove Windows startup", "Stops ReplayKit from launching OBS when Windows signs in.", lambda: configure_obs_startup(False, log=progress.log))
    add("Remove Windows sleep override", "Restores default Windows sleep behavior for OBS.", lambda: remove_sleep_override(log=progress.log))
    add("Remove OBS plugins", "Deletes ReplayKit OBS plugins from the OBS install folder.", lambda: remove_replaykit_plugins(log=progress.log))
    add("Remove OBS Stream Audio", "Uninstalls the ReplayKit virtual audio device.", lambda: uninstall_vbcable(log=progress.log))
    add("Remove virtual display driver", "Deletes the optional virtual display driver if it exists.", lambda: remove_virtual_display_driver(log=progress.log))
    if keep_user_settings:
        add("Keep ReplayKit settings", "Saves current ReplayKit settings for the next install.", lambda: save_user_settings(log=progress.log))
    else:
        add("Remove ReplayKit settings", "Deletes saved ReplayKit preferences.", lambda: remove_user_settings(log=progress.log))
    add("Wipe OBS ReplayKit config", "Deletes ReplayKit's OBS config folder while preserving OBS scenes and profiles.", lambda: wipe_obs_config(log=progress.log))

    progress.total_steps = len(steps)
    from .cli import _run_apply_step
    for index, (title, detail, action) in enumerate(steps, 1):
        _run_apply_step(progress, index, title, detail, action)
    progress.render(progress.total_steps, "Cleanup complete", "OBS ReplayKit changes were removed.", state="done")
    return progress.issues

File: ReplayKitSetup/Cleanup.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Management;

namespace ReplayKitSetup
{
    // in-menu cleanup for OBS ReplayKit installs. ported from obs_replaykit/cleanup.py.
    public static class Cleanup
    {
        private static readonly string[] ReplaykitRuntimeStateRels =
        {
            "obs-replayKit/scripts/helper/clips_db.json",
            "obs-replayKit/scripts/helper/clips_index.json",
        };

        private static (int ExitCode, string Stdout, string Stderr) RunHidden(string fileName, string arguments, int timeoutMs = 30000)
        {
            var psi = new ProcessStartInfo(fileName, arguments)
            {
                UseShellExecute = false,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                CreateNoWindow = true,
            };
            using (var proc = Process.Start(psi))
            {
                string stdout = proc.StandardOutput.ReadToEnd();
                string stderr = proc.StandardError.ReadToEnd();
                if (!proc.WaitForExit(timeoutMs))
                {
                    try { proc.Kill(); } catch (InvalidOperationException) { }
                    throw new TimeoutException(fileName + " timed out");
                }
                return (proc.ExitCode, stdout, stderr);
            }
        }

        private static int? CurrentParentPid()
        {
            try
            {
                using (var searcher = new ManagementObjectSearcher($"SELECT ParentProcessId FROM Win32_Process WHERE ProcessId={Process.GetCurrentProcess().Id}"))
                using (var results = searcher.Get())
                {
                    foreach (ManagementObject row in results)
                    {
                        return Convert.ToInt32(row["ParentProcessId"]);
                    }
                }
                return null;
            }
            catch (ManagementException)
            {
                return null;
            }
        }

        // close obs and replaykit helper processes while keeping this setup process alive.
        public static bool StopObsAndHelpers(Action<string> log = null)
        {
            Obs.CloseObs(log);
            var keep = new HashSet<int> { Process.GetCurrentProcess().Id };
            var parent = CurrentParentPid();
            if (parent.HasValue) keep.Add(parent.Value);
            string keepCsv = string.Join(",", keep);
            string script =
                $"$keep = @({keepCsv})\r\n" +
                "Get-CimInstance Win32_Process |\r\n" +
                "  Where-Object { $_.Name -in @('OBSReplayKit.exe','OBSReplayKit-Encoder.exe') -and $keep -notcontains [int]$_.ProcessId } |\r\n" +
                "  ForEach-Object {\r\n" +
                "    try { Stop-Process -Id $_.ProcessId -Force -ErrorAction Stop } catch { }\r\n" +
                "  }\r\n";
            try
            {
                RunHidden("powershell.exe", Win32Args.Build("-NoProfile", "-NonInteractive", "-Command", script));
            }
            catch (Exception exc) when (exc is System.ComponentModel.Win32Exception || exc is TimeoutException)
            {
                log?.Invoke("warn: helper stop failed: " + exc.Message);
                return false;
            }
            return true;
        }

        public static bool RemoveReplaykitPlugins(Action<string> log = null)
        {
            string obsRoot = Config.PROGRAMFILES_OBS_DIR;
            var targets = new[]
            {
                Path.Combine(obsRoot, "obs-plugins", "64bit", "win-capture-audio.dll"),
                Path.Combine(obsRoot, "obs-plugins", "64bit", "win-capture-audio.pdb"),
                Path.Combine(obsRoot, "data", "obs-plugins", "win-capture-audio"),
                Path.Combine(obsRoot, "obs-plugins", "64bit", "input-overlay.dll"),
                Path.Combine(obsRoot, "obs-plugins", "64bit", "SDL2.dll"),
                Path.Combine(obsRoot, "data", "obs-plugins", "input-overlay"),
                Path.Combine(obsRoot, "obs-plugins", "64bit", "bongobs-cat.dll"),
                Path.Combine(obsRoot, "bin", "64bit", "Bango Cat"),
                Path.Combine(obsRoot, "data", "obs-plugins", "bongobs-cat"),
                Path.Combine(obsRoot, "obs-plugins", "64bit", "obs-composite-blur.dll"),
                Path.Combine(obsRoot, "obs-plugins", "64bit", "obs-composite-blur.pdb"),
                Path.Combine(obsRoot, "data", "obs-plugins", "obs-composite-blur"),
                Path.Combine(Config.PROGRAMDATA, "obs-studio", "plugins", "obs-composite-blur"),
                Path.Combine(obsRoot, "obs-plugins", "64bit", "obs-shaderfilter.dll"),
                Path.Combine(obsRoot, "obs-plugins", "64bit", "obs-shaderfilter.pdb"),
                Path.Combine(obsRoot, "data", "obs-plugins", "obs-shaderfilter"),
                Config.REPLAYKIT_TRAY_PLUGIN_DIR,
            };
            bool ok = true;
            foreach (var target in targets)
            {
                bool isDir = Directory.Exists(target);
                bool isFile = File.Exists(target);
                if (!isDir && !isFile) continue;
                try
                {
                    if (isDir) Directory.Delete(target, true);
                    else File.Delete(target);
                }
                catch (Exception exc) when (exc is IOException || exc is UnauthorizedAccessException)
                {
                    ok = false;
                    log?.Invoke("warn: could not remove " + target + ": " + exc.Message);
                }
            }
            return ok;
        }

        public static bool RemoveVirtualDisplayDriver(Action<string> log = null)
        {
            const string script = @"
$ErrorActionPreference = 'Continue'
$vddDevice = Get-PnpDevice -ErrorAction SilentlyContinue |
    Where-Object { $_.FriendlyName -eq 'Virtual Display Driver' } |
    Select-Object -First 1
if ($vddDevice) {
    Disable-PnpDevice -InstanceId $vddDevice.InstanceId -Confirm:$false -ErrorAction SilentlyContinue | Out-Null
    pnputil.exe /remove-device $vddDevice.InstanceId | Out-Null
}
$drivers = pnputil.exe /enum-drivers | Out-String
$matches = [regex]::Matches($drivers, ""Published Name:\s+(oem\d+\.inf)\s+Original Name:\s+MttVDD\.inf"", ""IgnoreCase"")
foreach ($m in $matches) {
    pnputil.exe /delete-driver $m.Groups[1].Value /uninstall /force | Out-Null
}
if (Test-Path ""C:\IddSampleDriver"") {
    Remove-Item -Recurse -Force ""C:\IddSampleDriver"" -ErrorAction SilentlyContinue
}
";
            (int ExitCode, string Stdout, string Stderr) result;
            try
            {
                result = RunHidden("powershell.exe", Win32Args.Build("-NoProfile", "-NonInteractive", "-Command", script), 60000);
            }
            catch (Exception exc) when (exc is System.ComponentModel.Win32Exception || exc is TimeoutException)
            {
                log?.Invoke("warn: virtual display cleanup failed: " + exc.Message);
                return false;
            }
            if (result.ExitCode != 0)
            {
                string message = !string.IsNullOrEmpty(result.Stderr) ? result.Stderr : result.Stdout;
                string firstLine = string.IsNullOrWhiteSpace(message) ? "" : message.Trim().Split('\n')[0];
                log?.Invoke("warn: virtual display cleanup returned " + result.ExitCode + ": " + firstLine);
            }
            return result.ExitCode == 0;
        }

        public static bool WipeObsConfig(Action<string> log = null)
        {
            if (!Directory.Exists(Config.REPLAYKIT_CONFIG)) return true;
            try
            {
                Directory.Delete(Config.REPLAYKIT_CONFIG, true);
                return true;
            }
            catch (Exception exc) when (exc is IOException || exc is UnauthorizedAccessException)
            {
                log?.Invoke("warn: could not wipe " + Config.REPLAYKIT_CONFIG + ": " + exc.Message);
                return false;
            }
        }

        public static bool SaveUserSettings(Action<string> log = null)
        {
            bool ok = true;
            try
            {
                var prefs = Prefs.LoadPrefs();
                prefs.Save();
                log?.Invoke("ReplayKit user settings kept -> " + Prefs.PREFS_FILE);
            }
            catch (Exception exc)
            {
                ok = false;
                log?.Invoke("warn: could not keep ReplayKit user settings: " + exc.Message);
            }

            try
            {
                Directory.CreateDirectory(Config.REPLAYKIT_USER_STATE_CACHE);
                int kept = 0;
                foreach (var rel in ReplaykitRuntimeStateRels)
                {
                    string source = Path.Combine(Config.OBS_CONFIG, rel.Replace('/', Path.DirectorySeparatorChar));
                    if (!File.Exists(source)) continue;
                    string target = Path.Combine(Config.REPLAYKIT_USER_STATE_CACHE, Path.GetFileName(rel));
                    File.Copy(source, target, true);
                    kept++;
                }
                if (kept > 0) log?.Invoke("ReplayKit clip state kept -> " + Config.REPLAYKIT_USER_STATE_CACHE);
            }
            catch (Exception exc) when (exc is IOException || exc is UnauthorizedAccessException)
            {
                ok = false;
                log?.Invoke("warn: could not keep ReplayKit clip state: " + exc.Message);
            }
            return ok;
        }

        public static bool RemoveUserSettings(Action<string> log = null)
        {
            var targets = new List<string>
            {
                Prefs.PREFS_FILE,
                Path.Combine(Config.REPLAYKIT_SETUP_CACHE, "prefs.json"),
                Path.Combine(Config.REPLAYKIT_SETUP_CACHE, "clips_state.json"),
            };

            bool ok = true;
            var seen = new HashSet<string>();
            foreach (var target in targets)
            {
                string key;
                try { key = Path.GetFullPath(target).ToLowerInvariant(); }
                catch (Exception ex) when (ex is ArgumentException || ex is NotSupportedException) { continue; }
                if (!seen.Add(key)) continue;
                if (!File.Exists(target)) continue;
                try
                {
                    File.Delete(target);
                }
                catch (Exception exc) when (exc is IOException || exc is UnauthorizedAccessException)
                {
                    ok = false;
                    log?.Invoke("warn: could not remove " + target + ": " + exc.Message);
                }
            }
            if (Directory.Exists(Config.REPLAYKIT_USER_STATE_CACHE))
            {
                try
                {
                    Directory.Delete(Config.REPLAYKIT_USER_STATE_CACHE, true);
                }
                catch (Exception exc) when (exc is IOException || exc is UnauthorizedAccessException)
                {
                    ok = false;
                    log?.Invoke("warn: could not remove " + Config.REPLAYKIT_USER_STATE_CACHE + ": " + exc.Message);
                }
            }
            return ok;
        }

        public static List<string> RunCleanup(IInstallProgress progress, bool keepUserSettings = true)
        {
            var steps = new List<(string Title, string Detail, Func<object> Action)>
            {
                ("Close OBS", "Stops OBS and ReplayKit helpers, but keeps this setup window alive.", () => StopObsAndHelpers(progress.LogLine)),
                ("Remove launch permission", "Deletes the ReplayKit scheduled task.", () => ScheduledTask.DeleteElevationTask(progress.LogLine)),
                ("Remove Windows startup", "Stops ReplayKit from launching OBS when Windows signs in.", () => Startup.ConfigureObsStartup(false, progress.LogLine)),
                ("Remove Windows sleep override", "Restores default Windows sleep behavior for OBS.", () => SleepOverride.RemoveSleepOverride(progress.LogLine)),
                ("Remove OBS plugins", "Deletes ReplayKit OBS plugins from the OBS install folder.", () => RemoveReplaykitPlugins(progress.LogLine)),
                ("Remove OBS Stream Audio", "Uninstalls the ReplayKit virtual audio device.", () => VbCable.UninstallVbcable(progress.LogLine)),
                ("Remove virtual display driver", "Deletes the optional virtual display driver if it exists.", () => RemoveVirtualDisplayDriver(progress.LogLine)),
            };
            if (keepUserSettings)
                steps.Add(("Keep ReplayKit settings", "Saves current ReplayKit settings for the next install.", () => SaveUserSettings(progress.LogLine)));
            else
                steps.Add(("Remove ReplayKit settings", "Deletes saved ReplayKit preferences.", () => RemoveUserSettings(progress.LogLine)));
            steps.Add(("Wipe OBS ReplayKit config", "Deletes ReplayKit's OBS config folder while preserving OBS scenes and profiles.", () => WipeObsConfig(progress.LogLine)));

            progress.TotalSteps = steps.Count;
            for (int i = 0; i < steps.Count; i++)
            {
                Cli.RunApplyStep(progress, i + 1, steps[i].Title, steps[i].Detail, steps[i].Action);
            }
            progress.Render(progress.TotalSteps, "Cleanup complete", "OBS ReplayKit changes were removed.", "done");
            return progress.Issues;
        }
    }
}
